utils/s2n_safety.h

```diff
@@ -43,6 +43,16 @@
  */
 #define BAIL_POSIX( x )                              do { _S2N_ERROR( ( x ) ); return S2N_FAILURE; } while (0)
 
+/**
+ * Sets the global `errno` and the crypto error. Then returns with a POSIX error (`-1`)
+ */
+#define BAIL_OSSL( x )                              do { _S2N_ERROR( ( x ) ); s2n_save_crypto_error(); return S2N_FAILURE; } while (0)
+
+/**
+ * Sets the global `errno` and the crypto error. Then a `NULL` pointer value
+ */
+#define BAIL_PTR_OSSL( x )                              do { _S2N_ERROR( ( x ) ); s2n_save_crypto_error(); return NULL; } while (0)
+
```

Comment:
My issue with this solution is it now requires even more safety macros (and there's still some missing for each permutation). Why not just have this be an inlined function instead. That way it's much easier to compose:

```c
inline S2N_RESULT ossl_capture(bool check)
{
    if (check) {
        return S2N_RESULT_OK;
    } else {
        s2n_libcrypto_error = ERR_peek_last_error();
        if (s2n_libcrypto_error != 0) {
            ERR_clear_error();
        }
        return S2N_RESULT_ERR;
    }
}

#define OSSL_RESULT(x) ossl_capture((x) == _OSSL_SUCCESS)
#define OSSL_PTR(x) ossl_capture((x) != NULL)
```

Now i can use those in whatever context I would like (POSIX, S2N_RESULT, PTR):

```c
GUARD_AS_POSIX(OSSL_RESULT(ossl_call_goes_here()));
GUARD_AS_POSIX(OSSL_PTR(my_ptr = ossl_call_goes_here()));
```

Reply:
Let me think for a few minutes on this, because I do rather like this solution.

My immediate concerns are:
* We may still need a proliferation of these to cover different inputs (though at least not return values)
* This makes it much harder (impossible?) to clear the libcrypto error state for non-openssl errors

I also tried this by simply capturing the libcrypto errors in all cases (see my [ossl_errors branch](https://github.com/awslabs/s2n/compare/main...SalusaSecondus:ossl_errors)). While much simpler, I'm concerned about the performance implications.

Reply:
> We may still need a proliferation of these to cover different inputs (though at least not return values)

Well at least it's just `N` rather than `N * M` :)

> This makes it much harder (impossible?) to clear the libcrypto error state for non-openssl errors

I'm not sure I totally understand the usecase for this. Could you provide an example?

Reply:
I'm saving the libcrypto error in a thread-local variable as well so that developers can interrogate it and we don't just drop it. (Maybe this is a misfeature and we should drop it, but that feels wrong.)
There are now two different thread local error states for our customers:
* `s2n_errno` (existing)
* `s2n_libcrypto_error` (new, and holds the most recent libcrypto error)

Now consider the following cases:

*My proposals*:
1. libcrypto error happens and is properly handled by the calller
  * `s2n_errno = S2N_ERROR_FOO`
  * `s2n_libcrypto_error = LIBCRYPTO_ERROR_BAR`
2. Non-libcrypto error happens. Now the error state is:
  * `s2n_errno = S2N_ERROR_BAZ`
  * `s2n_libcrypto_error = 0`
(This is correct)

What I meant when I (through very poor wording) said "clear the libcrypto error state for non-openssl errors" was specifically clearning the value of `s2n_libcrypto_error `. If we don't handle this correctly, the results of the above flow is as follows:
1. libcrypto error happens and is properly handled by the calller
  * `s2n_errno = S2N_ERROR_FOO`
  * `s2n_libcrypto_error = LIBCRYPTO_ERROR_BAR`
2. Non-libcrypto error happens. Now the error state is:
  * `s2n_errno = S2N_ERROR_BAZ`
  * `s2n_libcrypto_error = LIBCRYPTO_ERROR_BAR`
(This gives the wrong value for `s2n_libcrypto_error`)

Reply:
Ok I think I understand, now. But didn't you already take care of that in the `_S2N_ERROR` macro?

https://github.com/awslabs/s2n/blob/0cc779fa5dded0ea971e1f882e14414b0d183898/error/s2n_errno.h#L274

Reply:
Speaking of setting the `s2n_errno`, we'll probably need to either convert the libcrypto error code into a `s2n_errno` or add a generic `S2N_ERR_LIBCRYPTO` error that indicates callers should look at `s2n_libcrypto_error` for more details.

```c
/* conversion option */
inline S2N_RESULT ossl_ensure(bool check)
{
    if (check) {
        return S2N_RESULT_OK;
    } else {
        s2n_libcrypto_error = ERR_peek_last_error();
        if (s2n_libcrypto_error != 0) {
            s2n_errno = ossl_covert_errno_to_s2n(s2n_libcrypto_error);
            ERR_clear_error();
        }
        return S2N_RESULT_ERR;
    }
}
```

```c
/* generic error option */
inline S2N_RESULT ossl_ensure(bool check)
{
    if (check) {
        return S2N_RESULT_OK;
    } else {
        s2n_libcrypto_error = ERR_peek_last_error();
        if (s2n_libcrypto_error != 0) {
            s2n_errno = S2N_ERR_LIBCRYPTO;
            ERR_clear_error();
        }
        return S2N_RESULT_ERR;
    }
}
```

Reply:
1. I'd have to remove the `s2n_libcrypto_error = 0` from `_S2N_ERROR` with your solution because it would overwrite the value from `OSSL_RESULT`.
2. I don't think we should try to translate these error values. There are just too many of them and it is too complicated. However, if we were to create  a generic `S2N_ERR_LIBCRYPTO` value it would make everything easier. Then we could just rewrite `_S2N_ERROR` as follows:

```
#define _S2N_ERROR( x )     do { s2n_debug_str = _S2N_DEBUG_LINE; s2n_errno = ( x ); if ( s2n_errno == S2N_ERR_LIBCRYPTO ) { capture_libcrypto_error(); } s2n_libcrypto_error = 0; s2n_calculate_stacktrace(); } while (0) 
```

Callers would then signal that it is a libcrypto error by setting that as the error return value.

Reply:
> I'd have to remove the s2n_libcrypto_error = 0 from _S2N_ERROR with your solution because it would overwrite the value from OSSL_RESULT.

I don't think you would. Since we'd be using `GUARD`, it would just propagate both errnos.

> I don't think we should try to translate these error values. There are just too many of them and it is too complicated.

Agreed! Having a generic `S2N_ERR_LIBCRYPTO` would be my preference.

Reply:
You're correct. I didn't realize  that GUARD didn't route through `__S2N_ERROR` (as `BAIL` and `ENSURE` do).

How much would it break things to introduce a new error code? Also, do you think it is better to:
1. Pick path based on error code
2. Use the wrapper you designed as above
3. Both?

Reply:
> How much would it break things to introduce a new error code?

We do it all the time! Customers aren't able to read error codes currently: see #882

> Also, do you think it is better to:
>
> 1. Pick path based on error code
> 1. Use the wrapper you designed as above
> 1. Both?

I don't think we should do both (option 3). We should probably pick one. I'm not sure it actually matters, though. I guess putting it in the `_S2N_ERROR` does allow for any `ENSURE(some_check, S2N_ERR_LIBCRYPTO)` call to retrieve the error code, so I guess my preference would be 1.

The other question is what will the UI look like:

### Option 1

This option will most likely require `N` * `M` macros where `N` is the possible return types of s2n functions (`[POSIX, RESULT, PTR]`) and `M` is the possible return types of OpenSSL functions (`[RESULT, PTR]`).

```c
/* POSIX */
POSIX_OSSL_RESULT(ossl_call_goes_here());
POSIX_OSSL_PTR(my_ptr = ossl_call_goes_here());

/* S2N_RESULT */
RESULT_OSSL_RESULT(ossl_call_goes_here());
RESULT_OSSL_PTR(my_ptr = ossl_call_goes_here());

/* PTR */
PTR_OSSL_RESULT(ossl_call_goes_here());
PTR_OSSL_PTR(my_ptr = ossl_call_goes_here());
```

### Option 2 (my vote)

The nice thing about this option is we only need to convert OpenSSL return types (`[RESULT, PTR]`) into an `S2N_RESULT`, which only requires 2 new macros. This would also allow us to get rid of `GUARD_OSSL`, `GUARD_POSIX_OSSL`, `GUARD_RESULT_OSSL`

```c
/* POSIX */
GUARD_AS_POSIX(OSSL_RESULT(ossl_call_goes_here()));
GUARD_AS_POSIX(OSSL_PTR(my_ptr = ossl_call_goes_here()));

/* S2N_RESULT */
GUARD_RESULT(OSSL_RESULT(ossl_call_goes_here()));
GUARD_RESULT(OSSL_PTR(my_ptr = ossl_call_goes_here()));

/* PTR */
GUARD_RESULT_PTR(OSSL_RESULT(ossl_call_goes_here()));
GUARD_RESULT_PTR(OSSL_PTR(my_ptr = ossl_call_goes_here()));
```

### Option 3

No new macros at all. Pretty verbose... :-1: 

```c
/* POSIX */
ENSURE_POSIX(ossl_call_goes_here() == _OSSL_SUCCESS, S2N_ERR_LIBCRYPTO);
ENSURE_POSIX((my_ptr = ossl_call_goes_here()) != NULL, S2N_ERR_LIBCRYPTO);

/* etc */
```

Reply:
Hmm. So looking at the code that you've converted, I'm wondering what this would look like:

https://github.com/awslabs/s2n/blob/0cc779fa5dded0ea971e1f882e14414b0d183898/crypto/s2n_ecc_evp.c#L117

If `_S2N_ERROR` sets the `s2n_libcrypto_error` based on `s2n_errno`, then this wouldn't be possible.

I wonder if we go with my inline function proposal in addition to implementing #1941?

```c
GUARD_WITH(OSSL_RESULT(ossl_call_goes_here()), S2N_ERR_ECDHE_GEN_KEY);
```

This would result in 

*    s2n_errno = S2N_ERR_ECDHE_GEN_KEY
*    s2n_libcrypto_error = LIBCRYPTO_ERROR_BAR

And in the normal case where we don't specify anything more specific:

```c
GUARD(OSSL_RESULT(ossl_call_goes_here()));
```

This would result in 

*    s2n_errno = S2N_ERR_LIBCRYPTO
*    s2n_libcrypto_error = LIBCRYPTO_ERROR_BAR

Reply:
I'm still a bit worried about other flows leaving `s2n_libcrypto_error` untouched, but so long as they use something which routes through `__S2N_ERROR` or `OSSL_*` they should be fine.

Reply:
Oh you're totally right: not having the check/reset in `_S2N_ERROR` would mean that the `s2n_libcrypto_error` only gets modified in `OSSL_*` macros...

What if, instead, we do something like this:

```c
inline S2N_RESULT ossl_ensure(bool check, int errno)
{
    if (check) {
        return S2N_RESULT_OK;
    } else {
        /* set the initial errno to S2N_ERR_LIBCRYPTO in order for `_S2N_ERROR` to capture the libcrypto error */
        _S2N_ERROR(S2N_ERR_LIBCRYPTO);
        /* override the actual errno */
        s2n_errno = errno;
        return S2N_RESULT_ERR;
    }
}

#define OSSL_RESULT(x) OSSL_RESULT_WITH((x), S2N_ERR_LIBCRYPTO)
#define OSSL_PTR(x) OSSL_PTR_WITH((x), S2N_ERR_LIBCRYPTO)
#define OSSL_RESULT_WITH(x, err) ossl_ensure((x) == _OSSL_SUCCESS, (err))
#define OSSL_PTR_WITH(x, err) ossl_ensure((x) != NULL, (err))
```

That way we can preserve the error codes while still capturing the libcrypto error:

```c
GUARD_AS_POSIX(OSSL_PTR_WITH(pctx, S2N_ERR_ECDHE_GEN_KEY));
```